_posts/2023-07-22-gsoc-sqlancer-final.md

```diff
@@ -0,0 +1,71 @@
+q
+
+Currently, the AST-based transformations include:
+
++ Remove union selects.  e.g., 
+  `SELECT 1 UNION SELECT 2` -> `SELECT 1` 
++ Remove irrelevant clauses.  e.g., 
+  `SELECT * FROM t OFFSET 20 LIMIT 5` -> `SELECT * FROM t`
++ Remove list elements.  e.g.,
+   `SELECT a, b, c FROM t` -> `SELECT a FROM t`
++ Remove rows of an insert statement.  e.g., 
+  `INSERT INTO t VALUES (1, 2), (3, 4)` -> `INSERT INTO t VALUES (1, 2)`
++ Replace complicated expressions with their sub expressions.  e.g., 
+  `(a+b)+c` -> `c`
++ Simplify constant values.  e.g., 
+  `3.27842156` -> `3.278`
+
+The AST-based reducer is designed with extensibility. It walks through the list of transformations and applies them to statements until a fixed point is reached. Adding new transformations is straightforward. Simply create a new transformation and include it in the list. This flexibility allows for easy customization and expansion of the reducer's functionality. Additionally, it is also possible to support transformations that do not depend on JSQLParser.
+
+
+
+## Framework for reducers Testing
+
+Yutan designed a virtual database engine to facilitate the testing of reducers. Instead of executing the statements, this engine records them for analysis. One of its notable features is the ability to customize the conditions that trigger a bug. For instance, the testing framework allows specifying an interestingness check that tests whether certain words are part of the reduced SQL test case.
+
+This approach provides a testing environment and allows for thorough evaluation of the reducers' performance. It offers convenience and flexibility in refining and polishing the functionality of the reducers without impacting real databases.
+
+## Reduction logs
+
+If test-case reduction is enabled, each time the reducer performs a reduction step successfully,it prints the reduced statements to the log file, overwriting the previous ones.
+
+The log files will be stored in the following format: `logs/<DBMS>/reduce/<database>-reduce.log`. For instance, if the tested DBMS is SQLite3 and the current database is named database0, the log file will be located at `logs/sqlite3/reduce/database0-reduce.log`.
+
+## Usage
+Test-case reduction is disabled by default. The statement reducer can be enabled by passing `--use-reducer` when starting SQLancer. 
+
+If you wish to further shorten each statements, you need to additionally pass the `--reduce-ast` parameter so that the AST-based reduction is applied. 
+
+Note: if `--reduce-ast` is set, `--use-reducer` option must be enabled first.
+
+There are also options to define timeout seconds and max steps of reduction for both statement reducer and AST-based reducer.
+
+For more details, you can refer to this [doc](https://github.com/sqlancer/sqlancer/blob/main/docs/testCaseReduction.md).
```

Comment:
I think you could link to the current version of the document, in case it gets renamed in the future? The link is https://github.com/sqlancer/sqlancer/blob/7804a3adec0962ad6d24687c42ec473aa49669fe/docs/testCaseReduction.md.